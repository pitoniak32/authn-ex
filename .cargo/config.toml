[env]
OTEL_COLLECTOR_URI="grpc://localhost:4317"
DEV_MODE="true"

MONGO_DB_URI="mongodb://0.0.0.0:27017"
MONGO_DB_NAME="authn-ex"

