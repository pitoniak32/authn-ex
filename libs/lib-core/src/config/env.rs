// // Declare all the environment variables needed by the application.
// // Call a function to ensure all the required variables are present.
// // Quick access to the values based on keys

// use std::{collections::HashMap, env::VarError};

// use lazy_static::lazy_static;

// lazy_static! {
//     static ref ENVIRONMENT: HashMap<EnvKey, String> = {
//         let mut m = HashMap::new();
//         let result = load_env(&mut m);
//         m.insert(EnvKey::MongoUri, std::env::var("MONGO_URI").unwrap_or("TEST".to_string()));
//         m
//     };
//     static ref ENVIONMENT_COUNT: usize = ENVIRONMENT.len();
// }

// const ENV_KEY_MONGO_URI: &'static str = "MONGO_URI";
// const ENV_DEFAULT_MONGO_URI: &'static str = "mongodb://0.0.0.0:27017";

// #[derive(Debug, Clone, PartialEq, Eq, Hash)]
// pub enum EnvKey {
//   MongoUri,
// }

// fn load_env(map: &mut HashMap<EnvKey, String>) -> Vec<VarError> {
//   map.insert(EnvKey::MongoUri, std::env::var(ENV_KEY_MONGO_URI).unwrap_or(ENV_DEFAULT_MONGO_URI.to_owned()));
//   vec![]
// }

// #[cfg(test)]
// mod tests {
//     use rstest::rstest;
//     use similar_asserts::assert_eq;

//     use crate::config::env::{EnvKey, ENVIONMENT_COUNT, ENVIRONMENT};

//     #[rstest]
//     fn test_lazy_environment() {
//         // Arrange
//         // Act
//         let mongo_uri = EnvKey::MongoUri.get();

//         // Assert
//         assert_eq!(mongo_uri, "TEST".to_string());
//     }
// }
