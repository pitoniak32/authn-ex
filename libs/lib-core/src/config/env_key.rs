use std::{
    env::{self},
    net::Ipv4Addr,
    str::FromStr,
};

#[derive(Debug, Clone)]
pub enum EnvKey {
    /// Uri used to configure otlp service.
    OtelCollectorUri,
    /// Uri used to connect to mongodb.
    MongoDbUri,
    /// Name used to create db in mongodb.
    MongoDbName,
    /// Port that the service will bind to.
    ServicePort,
    /// Ip that the service will bind to.
    ServiceIp,
    /// Determines if the app should be configured for development, or production.
    DevMode,
}

pub const APP_NAME: &str = env!("CARGO_PKG_NAME");
pub const DEFAULT_SERVICE_PORT: u16 = 8080;
pub const DEFAULT_SERVICE_IP: Ipv4Addr = Ipv4Addr::new(0, 0, 0, 0);
pub const DEFAULT_OTEL_URI: &str = "https://0.0.0.0:4317";
pub const DEFAULT_MONGO_URI: &str = "mongodb://0.0.0.0:27017";
pub const DEFAULT_LOG_FILTER: &str = "INFO";
pub const DEFAULT_DEV_MODE: bool = false;
pub const AUTH_TOKEN_STRING: &str = "access_token";

pub const ENV_KEY_SERVICE_PORT: &str = "SERVICE_PORT";
pub const ENV_KEY_SERVICE_IP: &str = "SERVICE_IP";
pub const ENV_KEY_OTEL_COLLECTOR_URI: &str = "OTEL_COLLECTOR_URI";
pub const ENV_KEY_MONGO_DB_URI: &str = "MONGO_DB_URI";
pub const ENV_KEY_MONGO_DB_NAME: &str = "MONGO_DB_NAME";
pub const ENV_KEY_DEV_MODE: &str = "DEV_MODE";

pub trait AsStr {
    fn as_str(&self) -> &'static str;
}

impl AsStr for EnvKey {
    fn as_str(&self) -> &'static str {
        match self {
            EnvKey::ServicePort => ENV_KEY_SERVICE_PORT,
            EnvKey::ServiceIp => ENV_KEY_SERVICE_IP,
            EnvKey::OtelCollectorUri => ENV_KEY_OTEL_COLLECTOR_URI,
            EnvKey::MongoDbUri => ENV_KEY_MONGO_DB_URI,
            EnvKey::MongoDbName => ENV_KEY_MONGO_DB_NAME,
            EnvKey::DevMode => ENV_KEY_DEV_MODE,
        }
    }
}

use std::sync::OnceLock;

pub fn get_config() -> &'static Config {
    static INSTANCE: OnceLock<Config> = OnceLock::new();

    INSTANCE.get_or_init(Config::from_env)
}

#[allow(non_snake_case)]
pub struct Config {
    pub SERVICE_PORT: u16,
    pub SERVICE_IP: Ipv4Addr,
    pub OTEL_COLLECTOR_URI: Option<String>,
    pub MONGO_DB_URI: String,
    pub MONGO_DB_NAME: String,
    pub DEV_MODE: bool,
}

impl Config {
    pub fn from_env() -> Self {
        let service_port = get_value(EnvKey::ServicePort, DEFAULT_SERVICE_PORT);
        let service_ip = get_value(EnvKey::ServiceIp, DEFAULT_SERVICE_IP);
        let otel_collector_uri = get_optional_value(EnvKey::OtelCollectorUri);
        let mongo_db_uri = get_value(EnvKey::MongoDbUri, DEFAULT_MONGO_URI.to_string());
        let mongo_db_name = get_value(EnvKey::MongoDbName, APP_NAME.to_string());
        let dev_mode = get_value(EnvKey::DevMode, DEFAULT_DEV_MODE);

        Config {
            SERVICE_PORT: service_port,
            SERVICE_IP: service_ip,
            OTEL_COLLECTOR_URI: otel_collector_uri,
            MONGO_DB_URI: mongo_db_uri,
            MONGO_DB_NAME: mongo_db_name,
            DEV_MODE: dev_mode,
        }
    }
}

fn get_value<T, R>(key: T, default: R) -> R
where
    T: AsStr,
    R: FromStr,
{
    match env::var(key.as_str()) {
      Ok(value) => value.parse::<R>().unwrap_or_else(|_| {
          panic!(
              "{} should be a valid {}! {} is not valid. To use default unset {} environment variable.",
              key.as_str(),
              std::any::type_name::<R>(),
              value,
              key.as_str(),
          )
      }),
      Err(_) => default,
  }
}

fn get_optional_value<T, R>(key: T) -> Option<R>
where
    T: AsStr,
    R: FromStr,
{
    match env::var(key.as_str()) {
      Ok(value) => Some(value.parse::<R>().unwrap_or_else(|_| {
          panic!(
              "{} should be a valid {}! {} is not valid. To use default unset {} environment variable.",
              key.as_str(),
              std::any::type_name::<R>(),
              value,
              key.as_str(),
          )
      })),
      Err(_) => None,
  }
}
