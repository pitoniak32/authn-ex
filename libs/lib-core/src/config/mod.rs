mod env;
mod env_key;
mod mongo;
mod otel;

pub use env_key::*;
pub use mongo::*;
pub use otel::*;
