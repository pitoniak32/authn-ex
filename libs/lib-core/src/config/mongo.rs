use mongodb::{Client, Collection};

use crate::model::{UserModel, UserSessionModel};

use super::get_config;

pub const MONGO_COLL_NAME_USER_SESSIONS: &str = "user_sessions";
pub const MONGO_COLL_NAME_USERS: &str = "users";

pub async fn init_mongo() -> mongodb::Client {
    // TODO: add way to timeout if mongo does not connect.
    let client = mongodb::Client::with_uri_str(&get_config().MONGO_DB_URI)
        .await
        .unwrap_or_else(|_| {
            panic!(
                "failed to connect to mongo with uri: {}",
                get_config().MONGO_DB_URI
            )
        });

    client
        .database(&get_config().MONGO_DB_NAME)
        .collection::<UserModel>(MONGO_COLL_NAME_USERS)
        .create_index(
            mongodb::IndexModel::builder()
                .keys(mongodb::bson::doc! { "username": 1 })
                .options(
                    mongodb::options::IndexOptions::builder()
                        .unique(true)
                        .build(),
                )
                .build(),
        )
        .await
        .unwrap_or_else(|e| {
            dbg!(&e);
            panic!("creating database index for {MONGO_COLL_NAME_USERS} should succeed")
        });

    client
        .database(&get_config().MONGO_DB_NAME)
        .collection::<UserSessionModel>(MONGO_COLL_NAME_USER_SESSIONS)
        .create_index(
            mongodb::IndexModel::builder()
                .keys(mongodb::bson::doc! { "user_id": 1 })
                .options(
                    mongodb::options::IndexOptions::builder()
                        .unique(false)
                        .build(),
                )
                .build(),
        )
        .await
        .unwrap_or_else(|e| {
            dbg!(&e);
            panic!("creating database index for {MONGO_COLL_NAME_USER_SESSIONS} should succeed")
        });

    client
}

pub fn get_users_collection(client: &Client) -> Collection<UserModel> {
    client
        .database(&get_config().MONGO_DB_NAME)
        .collection(MONGO_COLL_NAME_USERS)
}

pub fn get_sessions_collection(client: &Client) -> Collection<UserSessionModel> {
    client
        .database(&get_config().MONGO_DB_NAME)
        .collection(MONGO_COLL_NAME_USER_SESSIONS)
}
