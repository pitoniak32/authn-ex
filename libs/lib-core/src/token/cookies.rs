use tower_cookies::{cookie::time::Duration, Cookie};

use crate::config::get_config;

pub const COOKIE_ACCESS_TOKEN_KEY: &str = "ACCESS_TOKEN";
pub const COOKIE_REFRESH_TOKEN_KEY: &str = "REFRESH_TOKEN";
pub const COOKIE_TOKEN_PATH: &str = "/";
pub const COOKIE_HTTP_ONLY: bool = true;
pub const COOKIE_MAX_AGE: Duration = Duration::seconds(999999);

pub fn get_token_cookie<'r>(token: impl Into<String>, key: &'static str) -> Cookie<'r> {
    let mut cookie = Cookie::new(key, token.into());
    if !get_config().DEV_MODE {
        cookie.set_secure(true)
    }
    cookie.set_path(COOKIE_TOKEN_PATH);
    cookie.set_http_only(COOKIE_HTTP_ONLY);
    cookie.set_max_age(COOKIE_MAX_AGE);
    cookie
}

#[cfg(test)]
mod tests {
    use super::get_token_cookie;

    use rstest::rstest;

    #[rstest]
    async fn test_get_token_cookie() {
        // Act
        let cookie = get_token_cookie("test_get_token_cookie ok", "TEST_ACCESS_TOKEN_KEY");

        // Assert
        insta::assert_debug_snapshot!(cookie);
    }
}
