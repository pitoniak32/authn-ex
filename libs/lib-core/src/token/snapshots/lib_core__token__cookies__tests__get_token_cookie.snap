---
source: libs/lib-core/src/token/cookies.rs
expression: cookie
---
Cookie {
    cookie_string: None,
    name: Concrete(
        "TEST_ACCESS_TOKEN_KEY",
    ),
    value: Concrete(
        "test_get_token_cookie ok",
    ),
    expires: None,
    max_age: Some(
        Duration {
            seconds: 999999,
            nanoseconds: 0,
        },
    ),
    domain: None,
    path: Some(
        Concrete(
            "/",
        ),
    ),
    secure: None,
    http_only: Some(
        true,
    ),
    same_site: None,
    partitioned: None,
}
